### PR TITLE
receive: Optimized receive; fixed long term overallocation

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/thanos-io/thanos/issues/3265

## Changes:

* Added benchmarks
* Leveraged AddFast and removed debug prints on each sample
* Fixed over alloc by using normal Labels for remote write. After benchmarks, it's clear it does not give much.
* Optimized hash function.

Overall before vs after:
```
✗ benchstat -delta-test=none _dev/diffa.out _dev/difff.out
name                                                                                   old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                         21.7µs ± 0%    20.5µs ± 0%    -5.52%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12            35.0µs ± 0%    30.7µs ± 0%   -12.20%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                        72.4ms ± 0%    31.9ms ± 0%   -55.94%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12            1.09s ± 0%     0.02s ± 0%   -98.61%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                           26.5µs ± 0%    38.0µs ± 0%   +43.61%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12              38.0µs ± 0%    27.6µs ± 0%   -27.44%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                           140ms ± 0%      28ms ± 0%   -80.27%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12              1.30s ± 0%     0.01s ± 0%   -98.93%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                 18.4ms ± 0%    16.7ms ± 0%    -8.97%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12     162ms ± 0%      15ms ± 0%   -90.71%

name                                                                                   old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                         6.70kB ± 0%    6.32kB ± 0%    -5.76%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12            15.7kB ± 0%     9.2kB ± 0%   -41.22%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                        17.4MB ± 0%    15.3MB ± 0%   -12.22%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12            639MB ± 0%      21MB ± 0%   -96.71%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                           10.9kB ± 0%     8.8kB ± 0%   -19.57%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12              20.4kB ± 0%    11.7kB ± 0%   -42.73%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                           228MB ± 0%      16MB ± 0%   -93.07%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12              895MB ± 0%      21MB ± 0%   -97.65%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                 32.5MB ± 0%    33.0MB ± 0%    +1.33%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12    79.8MB ± 0%    33.0MB ± 0%   -58.66%

name                                                                                   old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                           48.0 ± 0%      66.0 ± 0%   +37.50%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12               126 ± 0%       106 ± 0%   -15.87%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                         5.12k ± 0%     5.13k ± 0%    +0.23%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12            4.00M ± 0%     0.00M ± 0%  -100.00%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                             50.0 ± 0%      67.0 ± 0%   +34.00%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12                 127 ± 0%       108 ± 0%   -14.96%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                            105k ± 0%        5k ± 0%   -95.13%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12              4.10M ± 0%     0.00M ± 0%  -100.00%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                   55.0 ± 0%      46.0 ± 0%   -16.36%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12       119 ± 0%        87 ± 0%   -26.89%
```
NOTE that `HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_of_samples/OK-12` and error, were exploding (too slow), so can only compare with Add Fast method:
```
 HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/OK-12                    5.10k ± 0%     5.10k ± 0%   +0.02%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/conflict_errors-12        500k ± 0%        0k ± 0%  -99.97%
```

### CODE REVIEW

**Comment by bwplotka** [comment]
Probably we can find more optimization, so far I can see this (inuse):
image

(allocs):
image

File: pkg/receive/handler.go [hunk 1]
```diff
@@ -314,6 +319,10 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 	if len(tenant) == 0 {
 		tenant = h.options.DefaultTenantID
 	}
+	if len(tenant) == 0 {
+		http.Error(w, "no tenant ID supplied", http.StatusBadRequest)
+		return
+	}
```
Comment (yeya24) [inline pkg/receive/handler.go:325]: What's the difference between this if and line 319?

File: pkg/receive/handler_test.go [hunk 2]
```diff
@@ -19,12 +23,16 @@ import (
...
 	"github.com/golang/snappy"
 	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/prometheus/pkg/labels"
 	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
 	terrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/testutil"
```
Comment (yeya24) [inline pkg/receive/handler_test.go:33]: Please group order these imports.

File: pkg/receive/multitsdb_test.go [hunk 3]
```diff
@@ -16,10 +16,10 @@ import (
 	"github.com/prometheus/prometheus/pkg/labels"
 	"github.com/prometheus/prometheus/storage"
 	"github.com/prometheus/prometheus/tsdb"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
 	"golang.org/x/sync/errgroup"
 
 	"github.com/thanos-io/thanos/pkg/runutil"
-	"github.com/thanos-io/thanos/pkg/store/labelpb"
```
Comment (yeya24) [inline pkg/receive/multitsdb_test.go:22]: It is not necessary to change this.

File: pkg/receive/handler.go [hunk 4]
```diff
@@ -282,13 +283,17 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 	span, ctx := tracing.StartSpan(r.Context(), "receive_http")
 	defer span.Finish()
 
-	compressed, err := ioutil.ReadAll(r.Body)
+	compressed := &bytes.Buffer{}
+	if r.ContentLength >= 0 {
+		compressed.Grow(int(r.ContentLength))
+	}
+	_, err := io.Copy(compressed, r.Body)
```
Comment (yeya24) [inline pkg/receive/handler.go:290]: Is this necessary? I am curious how much does it optimize?

File: pkg/receive/handler.go [hunk 5]
```diff
@@ -524,7 +534,7 @@ func (h *Handler) fanoutForward(pctx context.Context, tenant string, replicas ma
 							b.attempt++
 							dur := h.expBackoff.ForAttempt(b.attempt)
 							b.nextAllowed = time.Now().Add(dur)
-							level.Debug(h.logger).Log("msg", "target unavailable backing off", "for", dur)
+							level.Debug(h.logger).Log(append(logTags, "msg", "msg", "target unavailable backing off", "for", dur)...)
```
Comment (kakkoyun) [inline pkg/receive/handler.go:537]:
```suggestion
							level.Debug(h.logger).Log(append(logTags, "msg", "target unavailable backing off", "for", dur)...)
```

File: pkg/receive/handler.go (hunk 4 above)
Comment (bwplotka) [reply pkg/receive/handler.go:290]: I will comment (:

**Pull request closed without merging**